<link href="/Users/si/markdown/style.css" rel="stylesheet">

# 「if文のelif」&「try-except」構文で例外処理のパターンを知る

## if文って？？  
**if文**は、プログラムの中で**条件分岐**したいときに使ってね。  
「もし……だったら？」「もしくは……？」「どっちも違う？ そしたら……かな」みたいな感じ。

### elifとは  
**elif**とは、if文に書かれた条件と一致しなかった場合に、別の条件と一致しているかを確認するための構文。

### 順番が好き！
if、elif、elseは、同じ階層（インデント）で並べるよ！  
もしかしたら、順番が好きなのかもしれない。
だから、突然あっちこっちに飛ぶことはできないと思うからそこんとこは気をつけて！

#### 例えばこんな感じ
```
if 条件A:  
   処理A  
elif 条件B:  
   処理B  
else:　（空とか）
   処理C  
```
#### コードに直してみる
```
if a == "a":  
 処理A

elif a == "b":  
 処理B

else:  
 処理C  
```
#### 実際に使ってみる  
- 「変数**a**が10以上」であれば、**a**を2倍した値を表示  
- 「変数**a**が-10以下」であれば、**a**を-3倍した値を表示  
- 「それ以外」であれば、**a**をそのまま表示  
```
def check_value(a):  
  if a >= 10:  
        print(a * 2)  
    elif a <= -10:  
        print(a * -3)  
    else:  
        print(a)  
a = -15
check_value(a) # => -45
```

`a = 15` で実行すると、まず `if a >= 10` のチェックを行う。  
一致しなかったから、`elif a <= -10` でチェックしてみたら、一致した！ そこで、`-3`倍にした値を出力したよ！　ってコードかな。  


#### フローチャート 
if文の概念を図にするとこんな感じ！  

<img src="patternA.png" width="60%">  

if：「条件A」に一致する場合は、「処理A」になる  
elif：「条件A」には一致しないが、条件Bには一致する場合は、「処理B」になる  
else：「条件A, B」のどちらにも一致しない場合は、「処理C」にする  


## try-except構文 
順番？ それってそんなに大事ですか？　ボクなら例外を効率よく処理することができます。って言いそう。  
エラーを未然に防ぎたい時に使うのかな。

### 構文エラーとは
構文エラーとは、文法的な間違いによって発生するエラーのこと！　Pythonにおける文法ルールなどに従わないコードを実行したときに発生するんだって。  
文法的に正しいコードを書いても、実行時にエラーが発生することがありあるんだと。  
それを、**例外**というらしい。

### try-exceptの使い方  
例外が発生した時の対処に使うらしい。
```
try: 
    例外が発生するかもしれないが、実行したい処理。

except エラー名:
    例外発生時に行う処理  
```
まず、tryの中に指定された処理が実行！　例外が発生しなければ、except文はスキップ！  
でも、try実行中にexceptキーワードの後に指定したエラー名と一致する「例外」が発生すると、except文が実行される。  
そして、それ以降に記述された処理も実行！（たぶん）  

このように「except」ブロックで例外を処理をすることで、エラー発生箇所で処理が停止することはない、とうことらしい。

#### サンプルがあります  
```
x = 1
y = 0
 
try:
    z = x / y
except ZeroDivisionError:
    print('You can not do this operation!')
 
print('Hello world!')
```
出力すると……?  
```
You can not do this operation!
Hello world!
```
↑これね、try-except構文を使って、ゼロで割ったときに「You can not do this operation!」という文字列を出力させてるの。

結果は良好！　try‐exceptブロック外に記述されたHello World!という文字列の出力も、正常に行われてます。そして、、、  

この、try-except文を使わないと、エラーで処理が止まってしまうらしい。  （ZeroDivisionError）


##  try-except構文を使ってみる  
さっきの**else**との使い方の違いに注目！　　

```
 def func_a():  
 処理A 

 def func_b():    
 処理B

 func_dict = {"a": func_a, "b": func_b}  
 try:  
    func_dict[x]()  
    except keyError:  
 処理C  
```


#### フローチャート
try-except文の概念を図にするとこんな感じ！
<img src="patternB.png" width="60%">  

フローチャートだよ。  




## まとめ  
なんとなくだけど、elifの時は、「○○か◯○以外か」みたいなときに使うイメージ。  
try expect

おしまい
</link>